src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -238,28 +250,120 @@ private SNITCPHandle CreateTcpHandle(string fullServerName, long timerExpire, ob
+                {
+                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, SNICommon.InvalidConnStringError, e);
+                    return null;
+                }
+            }
+            else
             {
                 SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, SNICommon.InvalidConnStringError, string.Empty);
                 return null;
             }
+        }
+
+        /// <summary>
+        /// Sends CLNT_UCAST_INST request for given instance name to SQL Sever Browser, and receive SVR_RESP from the Browser.
+        /// </summary>
+        /// <param name="browserHostname">SQL Sever Browser hostname</param>
+        /// <param name="instanceName">instance name for CLNT_UCAST_INST request</param>
+        /// <returns>SVR_RESP packets from SQL Sever Browser</returns>
+        private static byte[] SendInstanceInfoRequest(string browserHostname, string instanceName)
+        {
+            Debug.Assert(!String.IsNullOrWhiteSpace(browserHostname));
+            Debug.Assert(!String.IsNullOrWhiteSpace(instanceName));
+
+            const byte ClntUcastInst = 0x04;
+            int byteCount = Encoding.ASCII.GetByteCount(instanceName);
+            byte[] requestPacket = new byte[byteCount + 1];
+            requestPacket[0] = ClntUcastInst;
+            Encoding.ASCII.GetBytes(instanceName, 0, instanceName.Length, requestPacket, 1);
+
+            byte[] responsePacket = null;
+            const int timeout = 1000;
+            using (UdpClient client = new UdpClient(AddressFamily.InterNetwork))
+            {
+                Task<int> sendTask = client.SendAsync(requestPacket, requestPacket.Length, browserHostname, SqlServerBrowserPort);
+                Task<UdpReceiveResult> receiveTask = null;
+                if (sendTask.Wait(timeout) && (receiveTask = client.ReceiveAsync()).Wait(timeout))
+                {
+                    responsePacket = receiveTask.Result.Buffer;
+                }
+            }
+
+            return responsePacket;
+        }
+
+        /// <summary>
+        /// Finds port number for given instance name.
+        /// </summary>
+        /// <param name="browserHostname">SQL Sever Browser hostname</param>
+        /// <param name="instanceName">instance name to find port number</param>
+        /// <returns>port number for given instance name</returns>
+        private static int GetPortByInstanceName(string browserHostname, string instanceName)
+        {
+            Debug.Assert(!String.IsNullOrWhiteSpace(browserHostname));
+            Debug.Assert(!String.IsNullOrWhiteSpace(instanceName));
+
+            byte[] responsePacket = SendInstanceInfoRequest(browserHostname, instanceName);
+            if (responsePacket == null || responsePacket.Length <= 3 || BitConverter.ToUInt16(responsePacket, 1) != responsePacket.Length - 3)
+            {
+                throw new SocketException();
```

Comment:
Is SocketException the right exception type here (and below)?  Regardless, seems like it'd be good to include a message (from resources) that describes the issue?